Json error messages

matajoh opened the pull request:
This PR does two main things:

- Add a context to all policy enforcement calls, so we can log errors appropriately
- Change the errors returned by policy to contain base64 encoded error objects instead of error message strings

darracott commented on pkg/securitypolicy/securitypolicyenforcer.go:
```diff
@@ -971,79 +975,79 @@ type ClosedDoorSecurityPolicyEnforcer struct {
 }
 
-func (ClosedDoorSecurityPolicyEnforcer) EnforceCreateContainerPolicy(_, _ string, _ []string, _ []string, _ string, _ []oci.Mount, _ bool, _ bool, _ IDName, _ []IDName, _ string, _ *oci.LinuxCapabilities, _ string) (EnvList, *oci.LinuxCapabilities, bool, error) {
+func (ClosedDoorSecurityPolicyEnforcer) EnforceCreateContainerPolicy(_ context.Context, _ string, _ string, _ []string, _ []string, _ string, _ []oci.Mount, _ bool, _ bool, _ IDName, _ []IDName, _ string, _ *oci.LinuxCapabilities, _ string) (EnvList, *oci.LinuxCapabilities, bool, error) {
```
Why have blank "_" names for args in some functions here but specify no names at all for args in other functions in this section?    (Asking just to improve my (very new) understanding of Go rather than because I think something's wrong)

matajoh replied:
Great question. The rule in Go is that you can either have names for everything, or names for nothing. In this case, I could actually remove the names as they aren't used like I did elsewhere and probably should for consistency.

darracott commented on pkg/securitypolicy/securitypolicy.go:
```diff
@@ -130,6 +130,23 @@ func MeasureSeccompProfile(seccomp *specs.LinuxSeccomp) (string, error) {
+func ExtractPolicyDecision(errorMessage string) (string, error) {
+	re := regexp.MustCompile(policyDecisionPattern)
+	matches := re.FindStringSubmatch(errorMessage)
+	if len(matches) != 2 {
```
In the expected case, is there definitely always only one  `policyDecision< (.*) >policyDecision` in errorMessage? (If so then happy with re.FindStringSubmatch meaning you expect `matches` to have two elements)

matajoh replied:
Yes, there should be only one. If there is more than one then something very weird has happened, and we want this to error out.

darracott commented on pkg/securitypolicy/securitypolicyenforcer_rego.go:
```diff
@@ -301,40 +305,32 @@ func errorString(errors interface{}) string {
+	base64EncodedErrorJson := base64.RawURLEncoding.EncodeToString(errorJson)
 
-	return fmt.Errorf("%s not allowed by policy. Errors: %v. Input: %s. Error Objects: %s", enforcementPoint, errors, string(inputJSON), string(matchesJSON))
+	return fmt.Errorf("policyDecision< %s >policyDecision", base64EncodedErrorJson)
```
I think this probably answers my earlier question about how many `policyDecision< (*.) >policyDecision` could appear in an error msg

anmaxvl commented on pkg/securitypolicy/securitypolicyenforcer_rego.go:
```diff
@@ -35,6 +37,11 @@ func init() {
 }
 
+const capabilitiesNilError = "capabilities object provided by the UVM to the policy engine is nil"
+const invalidError = "Security policy is not valid. Please check security policy or re-generate with tooling."
```
nit: don't capitalize error messages.

matajoh replied:
this is technically a message that goes into the JSON so I don't think Go naming conventions apply

anmaxvl commented on pkg/securitypolicy/securitypolicyenforcer_rego.go:
```diff
@@ -301,40 +316,53 @@ func errorString(errors interface{}) string {
 	}
 
-	defaultMessage := fmt.Errorf("%s not allowed by policy. Security policy is not valid. Please check security policy or re-generate with tooling. Input: %s", enforcementPoint, string(inputJSON))
+	log.G(ctx).WithField("error", string(errorJson)).Debug("Policy decision")
```
can the error json contain any sensitive data? if so, should we also redact it somehow?

matajoh replied:
Shouldn't be a concern here, as only the input has sensitive data in it.

anmaxvl commented on pkg/securitypolicy/securitypolicyenforcer_rego.go:
```diff
@@ -301,40 +316,53 @@ func errorString(errors interface{}) string {
-	inputJSON, err := json.Marshal(policy.redactSensitiveData(input))
+func (policy *regoEnforcer) denyWithError(ctx context.Context, policyError error, input inputData) error {
+	policyDecision := map[string]interface{}{
+		"input":       input,
```
should we redact sensitive data here?

matajoh replied:
Yes, we absolutely need to redact sensitive data.

anmaxvl reviewed (commented):
a few comments other than that looks good.

DomAyre reviewed (approved):
Aside from outstanding comments, looks okay to my relatively untrained eye

matajoh committed:
Moving to structured JSON policy decisions.

This commit makes two major changes:

1. All policy enforcement points now receive a context objects and use it
   to log policy errors and denial decisions.
2. Policy denials are now conveyed as structured JSON objects.

Whereas previously policy denial was surfaced as a text error message,
the policy now generates a bracketed base64 encoded string:

policydecision< (base64) >policydecision

When decoded, this will be a JSON object with the following structure:

```json
{
  "input": <redacted input object>,
  "decision": "deny",
  "reason": <reason object explaining decision>,
  "policyError": <optional error field>
}
```

NB: the `"policyError"` field above is only present if the denial was
triggered by an actual error in the Rego policy.

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>
